### PR TITLE
AKS CRG for Aks v20211101preview

### PR DESCRIPTION
**Pull request by tonche**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
aks

File: src/aks-preview/azext_aks_preview/decorator.py [hunk 1]
```diff
@@ -1516,6 +1516,15 @@ def get_oidc_issuer_profile(self) -> ManagedClusterOIDCIssuerProfile:
 
         return profile
 
+    def get_crg_id(self) -> str:
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/decorator.py:1519]: I would appreciate it if you could also write a test case for this "getter" function.

**Review by wangzelin007 — COMMENTED** [review]
If you want to release the new extension version, please update setup.py.

File: src/aks-preview/azext_aks_preview/decorator.py (hunk 1 above)
Comment (tonche) [reply src/aks-preview/azext_aks_preview/decorator.py:1519]: Added a test case for it.

**Comment by tonche** [comment]
> 

Modified version in the setup.py

**Comment by tonche** [comment]
Closed by a mistake.

File: src/aks-preview/azext_aks_preview/_params.py [hunk 2]
```diff
@@ -160,6 +160,7 @@ def load_arguments(self, _):
         c.argument('workload_runtime', arg_type=get_enum_type(workload_runtimes), default=CONST_WORKLOAD_RUNTIME_OCI_CONTAINER)
         c.argument('snapshot_id', type=str, validator=validate_snapshot_id, is_preview=True)
         c.argument('enable_oidc_issuer', action='store_true', is_preview=True)
+        c.argument('crg_id', type=str, validator=validate_crg_id, is_preview=True)
```
Comment (zhoxing-ms) [inline src/aks-preview/azext_aks_preview/_params.py:163]: `type=str` is the default setting, so we don't need an explicit declaration

File: src/aks-preview/azext_aks_preview/_params.py [hunk 3]
```diff
@@ -263,6 +264,7 @@ def load_arguments(self, _):
             c.argument('workload_runtime', arg_type=get_enum_type(workload_runtimes), default=CONST_WORKLOAD_RUNTIME_OCI_CONTAINER)
             c.argument('gpu_instance_profile', arg_type=get_enum_type(gpu_instance_profiles))
             c.argument('snapshot_id', type=str, validator=validate_snapshot_id, is_preview=True)
+            c.argument('crg_id', type=str, validator=validate_crg_id, is_preview=True)
```
Comment (zhoxing-ms) [inline src/aks-preview/azext_aks_preview/_params.py:267]: Same as above

**Comment by tonche** [comment]
Removed, thanks.

Regards,
Tong


From: Xing Zhou ***@***.***>
Reply-To: Azure/azure-cli-extensions ***@***.***>
Date: Sunday, January 9, 2022 at 22:56
To: Azure/azure-cli-extensions ***@***.***>
Cc: Tong Chen ***@***.***>, State change ***@***.***>
Subject: Re: [Azure/azure-cli-extensions] CRG for Aks v20211101preview (PR #4292)


@zhoxing-ms commented on this pull request.

________________________________

In src/aks-preview/azext_aks_preview/_params.py<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-cli-extensions%2Fpull%2F4292%23discussion_r780925452&data=04%7C01%7CTong.Chen%40microsoft.com%7Cee4804ddf7a546937b3a08d9d4065322%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637773945909183323%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=ckkgyvq0NC8bfbbyXi1inor2R7KUR65tQnvkEqjbr5I%3D&reserved=0>:

> @@ -263,6 +264,7 @@ def load_arguments(self, _):

             c.argument('workload_runtime', arg_type=get_enum_type(workload_runtimes), default=CONST_WORKLOAD_RUNTIME_OCI_CONTAINER)

             c.argument('gpu_instance_profile', arg_type=get_enum_type(gpu_instance_profiles))

             c.argument('snapshot_id', type=str, validator=validate_snapshot_id, is_preview=True)

+            c.argument('crg_id', type=str, validator=validate_crg_id, is_preview=True)

Same as above

—
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-cli-extensions%2Fpull%2F4292%23pullrequestreview-847423767&data=04%7C01%7CTong.Chen%40microsoft.com%7Cee4804ddf7a546937b3a08d9d4065322%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637773945909233323%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=%2BfyRw7cMu%2F%2FOcaWnHgn50EusiZYbfylIJNetzqaZ50M%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKX6GYZHOD2B7DYRX5HQSIDUVJ7JXANCNFSM5LIJJLUQ&data=04%7C01%7CTong.Chen%40microsoft.com%7Cee4804ddf7a546937b3a08d9d4065322%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637773945909233323%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=TKKfZIM63%2FpBqcRGQ61CblwEN57H%2FhshebDyCZ14etY%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7CTong.Chen%40microsoft.com%7Cee4804ddf7a546937b3a08d9d4065322%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637773945909233323%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=xrqZVbpGfNfOxQUo58HxRGZvBMZbR%2FLnkD%2FJAbke%2B4U%3D&reserved=0> or Android<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7CTong.Chen%40microsoft.com%7Cee4804ddf7a546937b3a08d9d4065322%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637773945909233323%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=n9rVxrULpvDBAoHR8OF9GA3%2FYAxG3imXs%2FNvbmHsfNg%3D&reserved=0>.

File: src/aks-preview/azext_aks_preview/_params.py (hunk 2 above)
Comment (zhoxing-ms) [reply src/aks-preview/azext_aks_preview/_params.py]: @tonche Could you please address this comment?

File: src/aks-preview/azext_aks_preview/_help.py [hunk 4]
```diff
@@ -453,6 +456,8 @@
           text: az aks create -g MyResourceGroup -n MyManagedCluster --load-balancer-sku Standard --network-plugin azure --windows-admin-username azure --windows-admin-password 'replacePassword1234$' --enable-windows-gmsa --gmsa-dns-server "10.240.0.4" --gmsa-root-domain-name "contoso.com"
         - name: create a kubernetes cluster with a snapshot id.
           text: az aks create -g MyResourceGroup -n MyManagedCluster --kubernetes-version 1.20.9 --snapshot-id "/subscriptions/00000/resourceGroups/AnotherResourceGroup/providers/Microsoft.ContainerService/snapshots/mysnapshot1"
+        - name: create a kubernetes cluster with a Capacity Reservation Group(CRG) ID.
+          text: az aks create -g MyResourceGroup -n MyManagedCluster --kubernetes-version 1.20.9 --crg-id "/subscriptions/<SubID>/resourceGroups/<ResourceGroupName>/providers/Microsoft.ContainerService/CapacityReservationGroups/<MyCRGID>"
```
[the remark this answers is not in the document]
Comment (zhoxing-ms) [reply src/aks-preview/azext_aks_preview/_help.py:460]: @tonche Could you please address this comment?

[the remark this answers is not in the document]
Comment (tonche) [reply src/aks-preview/azext_aks_preview/_help.py:460]: For some technical reasons, we were working on E2E test with Postman recently. I will work on this CLI E2E soon.

[the remark this answers is not in the document]
Comment (tonche) [reply src/aks-preview/azext_aks_preview/_help.py]: I have added a test case for CRG and uploaded the recording yaml file now.

**Review by FumingZhang — APPROVED** [review]
LGTM :shipit:

**Pull request merged**